File: python/question2/question2.py
#!/usr/bin/env python

import fileinput
import sys
import operator

class frequency:
    def __init__(self):
        self.terms = dict()

    def add(self, term):
        if term in self.terms:
            self.terms[term] += 1
        else:
            self.terms[term] = 1

    def top(self, k):
        a = []
        # get terms into array
        for key in self.terms:
            a.append([key, self.terms[key]])
        # sort array pairs by value decending order
        a.sort(key = operator.itemgetter(1), reverse=True)
        # print k terms
        for term in a:
            print term[0]
            k -= 1
            if k == 0:
                break

freq = frequency()

n = int(sys.stdin.readline())

# read n terms line by line from stdin
for _ in xrange(n):
    term = sys.stdin.readline().rstrip('\n')
    freq.add(term)

k = int(sys.stdin.readline())

freq.top(k)

File: ruby/question2/question2.rb
#!/usr/bin/env ruby

class Frequency
  def initialize
    @terms = {}
  end

  def add(term)
    if @terms[term]
      @terms[term] += 1
    else
      @terms[term] = 1
    end
  end

  def top(k)
    a = @terms.to_a

    # sort decending by value
    a.sort! { |x,y| y[1] <=> x[1] }

    a.each do |key, value|
      puts "#{key}"
      k -= 1
      break if k == 0
    end
  end
end

freq = Frequency.new

count = ARGF.readline.to_i

count.times do
  term = ARGF.readline.gsub!("\n", "")
  freq.add(term)

  exit 1 if ARGF.eof?
end

k = ARGF.readline.to_i

freq.top(k)
